***
# Introduction
**Scripts used to process and analyze micro-C data; Yang et al**

## Packages
```{bash}
bwa/0.7.17
samtools/1.9
Juicer Tools/1.22.01
Stripenn/1.1.50
pairtools/0.3.0
fanc/0.9.17
mustache/1.0.1
```

## Sample names
```{bash}
sample='Y1Y2Y3Y4 O1O2O3O4'
```

## Project path
```{bash}
path="/path"
```


# Data pre-processing
```{bash}
#Index fasta file for reference genome (MM10)
samtools faidx mm10.fa

#Create a genome file from indexed reference genome (prints the first two columns into a new file)
cut -f1,2 mm10.fa.fai > mm10.genome

#Create bwa index or direct bwa to the genome build (here mm10)
bwa index mm10.fa
```


# Alignment to mm10 genome
```{bash}
for i in $sample
do
  bwa mem -5SP -T0 mm10.fa ${path}/${i}_R1.fastq.gz ${path}/${i}_R2.fastq.gz -o ${path}/${i}_aligned.sam
done
```


# Identify ligation junctions
```{bash}
#Create pairedsams
for i in $sample
do
  pairtools parse --min-mapq 40 --walks-policy 5unique --max-inter-align-gap 30 --nproc-in 8 \
  --nproc-out 8 --chroms-path mm10.genome ${path}/${i}_aligned.sam > ${path}/${i}_parsed.pairsam

#Sort parsed pairs 
  pairtools sort --nproc 16 ${path}/${i}_parsed.pairsam > ${path}/${i}_sorted.pairsam

#Remove PCR duplicates 
  pairtools dedup --nproc-in 8 --nproc-out 8 --mark-dups --output-stats ${path}/${i}_stats.txt \
  --output ${path}/${i}_dedup.pairsam ${path}/${i}_sorted.pairsam
done
```

## Obtaining bam files
```{bash}
for i in $sample
do
#Create .pairs
  pairtools split --nproc-in 8 --nproc-out 8 --output-pairs ${path}/${i}_mapped.pairs \
  --output-sam ${path}/${i}_unsorted.bam ${path}/${i}_dedup.pairsam

#Create bam file
  samtools sort -o ${path}/${i}_mapped.PT.bam ${path}/${i}_unsorted.bam

#Index
  samtools index ${path}/${i}_mapped.PT.bam
done
```

## Library QC
```{bash}
for i in $sample
do
  python3 ./Micro-C/get_qc.py -p ${path}/${i}_stats.txt > ${path}/${i}_libraryQC.txt
done
```


# Hi-C contact maps
```{bash}
for i in $sample
do
  java -Xmx48000m  -Djava.awt.headless=true -jar ${path}/juicer_tools_1.22.01.jar \
  pre ${path}/${i}_mapped.pairs ${path}/${i}_contact_map.hic ${path}/mm10.genome
done
```


## Compartments calling
```{bash}
#calling compartments at 100kb resolution
#Contacts per billion normalization was first performed before calling compartments according to Rosencrance et al. 2020. Mol Cell. (PMID: 32243828)
for i in $sample
do
  fanc compartments -g ${path}/mm10.fa -v ${path}/${i}_compartment_norm_100kb.txt \
  ${path}/${i}_contact_map_inter_30.hic@100kb ${path}/${i}_compartment_norm_100kb.ab
done
```


## TADs calling
```{bash}
#calling TADs at 10kb resolution
for i in $sample
do
  java -jar ${path}/juicer_tools_1.22.01.jar arrowhead -r 10000 -k KR \
  ${path}/${i}_contact_map.hic ${path}/${i}_TADs_10Kb.bed
done
```


## Loops calling
```{bash}
#calling loops at 5kb resolution
chrname='chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chrX chrY'
for i in $chrname
do
  mustache -f ${path}/Y1Y2Y3Y4_contact_map_inter_30.hic -ch $i -r 5kb -o ${path}/${i}_Y1Y2Y3Y4_loops_5kb.tsv
  mustache -f ${path}/O1O2O3O4_contact_map_inter_30.hic -ch $i -r 5kb -o ${path}/${i}_O1O2O3O4_loops_5kb.tsv
done
```

